### PR TITLE
Add "dev" Composer keyword

### PR DESCRIPTION
**Pull request by GaryJones**
As per https://getcomposer.org/doc/04-schema.md#keywords by including "static analysis" as a keyword in the `composer.json` file, Composer 2.4.0-RC1 and later will prompt users if the package is installed with `composer require` instead of `composer require --dev`. See https://github.com/composer/composer/pull/10960 for more info.

### CODE REVIEW

**Review by simPod — DISMISSED** [review]

**Comment by greg0ire** [comment]
Should you send a PR to Composer to add the "coding standard" keyword to the list of keywords triggering that behavior? Static analysis and coding standards are not the same thing IMO

**Comment by GaryJones** [comment]
> Should you send a PR to Composer to add the "coding standard" keyword to the list of keywords triggering that behavior? Static analysis and coding standards are not the same thing IMO

It's a fair observation, though one that the author of PHPCS (the [GitHub repo](https://github.com/squizlabs/PHP_CodeSniffer) has a "static-analysis" tag), and many PHPCS-related packages disagree with.

PHPCS **does** do static analysis (literally, analysis of the static code through the tokeniser), with the intent that it can then be used to fix up inconsistencies in code styles and implementations. Of course, Psalm, PHPStan and others also do static analysis with the intent to find inconsistencies in used types to prevent bugs. While the intent of these tools is different, the underlying first step of "read the raw code" is the same (albeit with different implementations), hence both being different types of "static analysis".

If you still feel uncomfortable, then adding "dev" keyword instead would achieve the same benefit, though that seems less specific to me.

**Comment by lcobucci** [comment]
> These are: `dev`, `testing`, `static analysis`.

Maybe we can use `dev` and skip semantics?

**Comment by derrabus** [comment]
> Should you send a PR to Composer to add the "coding standard" keyword to the list of keywords triggering that behavior?

I think that would be reasonable, no matter how we decide here.

> PHPCS **does** do static analysis (literally, analysis of the static code through the tokeniser)

I agree. PHPCS belongs to the family of static analysis tools. Adding that tag would be justified.

> adding "dev" keyword instead would achieve the same benefit

Also fine for me.

**Review by derrabus — DISMISSED** [review]

**Comment by greg0ire** [comment]
> It's a fair observation, though one that the author of PHPCS (the [GitHub repo](https://github.com/squizlabs/PHP_CodeSniffer) has a "static-analysis" tag), and many PHPCS-related packages disagree 

I cannot find a discussion about this in #3655 , so it's possible they merged that PR without thinking too much about this.

> I agree. PHPCS belongs to the family of static analysis tools. Adding that tag would be justified.

According to https://en.wikipedia.org/wiki/Static_program_analysis , the human equivalent of a static analysis tool is ""program understanding", [program comprehension](https://en.wikipedia.org/wiki/Program_comprehension), or [code review](https://en.wikipedia.org/wiki/Code_review)". I don't think that's what phpcs rules do, or at least it's not what they should do.

There has been a similar discussion in https://github.com/doctrine/coding-standard/pull/289 and there seems a be a disagreement on this point exactly.

**Comment by derrabus** [comment]
> According to https://en.wikipedia.org/wiki/Static_program_analysis , the human equivalent of a static analysis tool is ""program understanding", [program comprehension](https://en.wikipedia.org/wiki/Program_comprehension), or [code review](https://en.wikipedia.org/wiki/Code_review)". I don't think that's what phpcs rules do, or at least it's not what they should do.

If you click on "code review", you see that one goal of code review is

> improve internal code quality and [maintainability](https://en.wikipedia.org/wiki/Maintainability) (readability, uniformity, understandability, etc.)

And this is as far as I can tell what we use PHPCS for. It does not replace a code review, but it helps a human reviewer to focus on those aspects of a review that cannot be done by a robot.

That being said, I believe we're nit-picking here. The goal of this PR was to opt-in to a new composer feature. And I think we should try to achieve that. I personally don't mind whether we get there by tagging this package with "static analysis" or a different term supported by Composer, as long as we get there. 🙂

File: composer.json
Comment (greg0ire) [inline composer.json]:
```suggestion
        "dev",
```

**Review by greg0ire — CHANGES_REQUESTED** [review]
I doubt readability, uniformity, understandability is the part the authors of the SA article wanted to hint at when they mentioned code review, especially since they prefixed it with "program understanding", [program comprehension](https://en.wikipedia.org/wiki/Program_comprehension). Since you are OK with using `dev` to achieve the goal, let's change to that.

**Comment by GaryJones** [comment]
The "Allow edit by maintainers" checkbox is active, so I'll let you folks make the final decision and merge updates and the whole PR when you're ready :-)

**Review by greg0ire — DISMISSED** [review]

**Review by derrabus — DISMISSED** [review]

**Comment by kukulich** [comment]
@greg0ire The commit message `Add "static analysis" Composer keyword` is strange now :)

**Comment by greg0ire** [comment]
Well spotted!

**Review by derrabus — APPROVED** [review]
I took the liberty to update the PR title as well. 🙃

**Review by greg0ire — APPROVED** [review]

**Review by ostrolucky — APPROVED** [review]

**Pull request merged**
